server/src/main/java/org/elasticsearch/indices/cluster/IndicesClusterStateService.java

```diff
@@ -390,78 +387,40 @@ private void removeIndices(final ClusterChangedEvent event) {
-                    // state may result in a new shard being initialized while having the same allocation id as the currently started shard.
-                    logger.debug("{} removing shard (not active, current {}, new {})", shardId, currentRoutingEntry, newShardRouting);
-                    indexService.removeShard(shardId.id(), "removing shard (stale copy)");
-                } else if (newShardRouting.primary() && currentRoutingEntry.primary() == false && newShardRouting.initializing()) {
-                    assert currentRoutingEntry.initializing() : currentRoutingEntry; // see above if clause
-                    // this can happen when cluster state batching batches activation of the shard, closing an index, reopening it
-                    // and assigning an initializing primary to this node
-                    logger.debug("{} removing shard (not active, current {}, new {})", shardId, currentRoutingEntry, newShardRouting);
-                    indexService.removeShard(shardId.id(), "removing shard (stale copy)");
+            } else {
+                // remove shards based on routing nodes (no deletion of data)
+                for (Shard shard : indexService) {
+                    ShardRouting currentRoutingEntry = shard.routingEntry();
+                    ShardId shardId = currentRoutingEntry.shardId();
+                    ShardRouting newShardRouting = localRoutingNode.getByShardId(shardId);
+                    if (newShardRouting == null) {
+                        // we can just remove the shard without cleaning it locally, since we will clean it in IndicesStore
+                        // once all shards are allocated
+                        logger.debug("{} removing shard (not allocated)", shardId);
+                        indexService.removeShard(shardId.id(), "removing shard (not allocated)");
+                    } else if (newShardRouting.isSameAllocation(currentRoutingEntry) == false) {
+                        logger.debug(
+                            "{} removing shard (stale allocation id, stale {}, new {})",
+                            shardId,
+                            currentRoutingEntry,
+                            newShardRouting
+                        );
+                        indexService.removeShard(shardId.id(), "removing shard (stale copy)");
+                    } else if (newShardRouting.initializing() && currentRoutingEntry.active()) {
+                        // this can happen if the node was isolated/gc-ed, rejoins the cluster and a new shard with the same allocation id
+                        // is assigned to it. Batch cluster state processing or if shard fetching completes before the node gets a new
+                        // cluster
+                        // state may result in a new shard being initialized while having the same allocation id as the currently started
+                        // shard.
```

Comment:
nit: formatting

```suggestion
                        // this can happen if the node was isolated/gc-ed, rejoins the cluster and a new shard with the same allocation id
                        // is assigned to it. Batch cluster state processing or if shard fetching completes before the node gets a new
                        // cluster state may result in a new shard being initialized while having the same allocation id as the currently
                        // started shard.
```